cns/fsnotify/fsnotify.go

```diff
@@ -0,0 +1,147 @@
+package fsnotify
+
+import (
+	"context"
+	"io/fs"
+	"os"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+type cnsclient interface {
+	ReleaseIPs(ctx context.Context, ipconfig cns.IPConfigsRequest) error
+}
+
+type Watcher struct {
+	CnsClient cnsclient
+}
+
+func WatchFs(w *Watcher, path string, logger *zap.Logger) error {
+	// Create new fs watcher.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		logger.Error("error creating watcher", zap.Error(err))
+	}
+	defer watcher.Close()
+
+	// Start listening for events.
+	logger.Info("listening for events from watcher")
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				logger.Info("received event", zap.String("event", event.Name))
+				if event.Has(fsnotify.Create) {
+					logger.Info("file created, triggering release", zap.String("event", event.Name))
+					cnsReleaseErr := w.releaseIP(event.Name)
+					if cnsReleaseErr != nil {
+						logger.Error("failed to release IP from CNS", zap.Error(cnsReleaseErr))
+					}
+					deleteErr := RemoveFile(event.Name, path)
+					if deleteErr != nil {
+						logger.Error("failed to remove file", zap.Error(err))
+					}
+				}
+				if event.Has(fsnotify.Remove) {
+					logger.Info("file deleted", zap.String("event", event.Name))
+				}
+			case watcherErr, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				logger.Error("watcher error", zap.Error(watcherErr))
+			}
+		}
+	}()
+
+	// Add directory where intended deletes are kept
+	err = os.MkdirAll(path, 0o755) //nolint
+	if err != nil {
+		logger.Error("error making directory", zap.Error(err))
+	}
+	err = watcher.Add(path)
+	if err != nil {
+		logger.Error("watcher add directory error", zap.Error(err))
+	}
+
+	// list the directory then call ReleaseIPs
+	logger.Info("listing directory deleteIDs")
+	dirContents, err := os.ReadDir(path)
+	if err != nil {
+		logger.Error("error reading deleteID directory", zap.Error(err))
+	} else {
+		for _, file := range dirContents {
+			logger.Info("file to be deleted", zap.String("name", file.Name()))
+			cnsReleaseErr := w.releaseIP(file.Name())
+			if cnsReleaseErr != nil {
+				logger.Error("failed to release IP from CNS", zap.Error(cnsReleaseErr))
+			}
+			err := RemoveFile(file.Name(), path)
+			if err != nil {
+				logger.Error("failed to remove file", zap.Error(err))
+			}
+		}
+	}
+
+	return nil
+}
```

Comment:
this feels like it should be a method on the `Watcher` instead of having it passed in.
I also thing it should be a blocking call (which the caller must start the goroutine for) so that it can propagate errors from the fs library. Here's how I would thread the context (written entirely on the GH UI so the syntax might not be perfect...).
```suggestion
type Watcher struct {
	client cnsclient
	path   string
	l      *zap.Logger
}

// WatchFS starts the filesystem watcher to handle async Pod deletes.
// Blocks until the context is closed; returns underlying fsnotify errors
// if something goes fatally wrong.
func WatchFS(ctx context.Context) error {
	// Create new fs watcher.
	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		logger.Error("error creating watcher", zap.Error(err))
	}
	defer watcher.Close()

	c, cancel := context.WithCancel(ctx)
	// Start listening for events.
	logger.Info("listening for events from watcher")
	go func() {
		defer cancel()
		for {
			select {
			case := <-ctx.Done():
				return
			case event, ok := <-watcher.Events:
				if !ok {
					return
				}
				logger.Info("received event", zap.String("event", event.Name))
				if event.Has(fsnotify.Create) {
					logger.Info("file created, triggering release", zap.String("event", event.Name))
					cnsReleaseErr := w.releaseIP(event.Name)
					if cnsReleaseErr != nil {
						logger.Error("failed to release IP from CNS", zap.Error(cnsReleaseErr))
					}
					deleteErr := RemoveFile(event.Name, path)
					if deleteErr != nil {
						logger.Error("failed to remove file", zap.Error(err))
					}
				}
				if event.Has(fsnotify.Remove) {
					logger.Info("file deleted", zap.String("event", event.Name))
				}
			case watcherErr, ok := <-watcher.Errors:
				if !ok {
					return
				}
				logger.Error("watcher error", zap.Error(watcherErr))
			}
		}
	}()

	// Add directory where intended deletes are kept
	err = os.MkdirAll(path, 0o755) //nolint
	if err != nil {
		logger.Error("error making directory", zap.Error(err))
	}
	err = watcher.Add(path)
	if err != nil {
		logger.Error("watcher add directory error", zap.Error(err))
	}

	// list the directory then call ReleaseIPs
	logger.Info("listing directory deleteIDs")
	dirContents, err := os.ReadDir(path)
	if err != nil {
		logger.Error("error reading deleteID directory", zap.Error(err))
	} else {
		for _, file := range dirContents {
			logger.Info("file to be deleted", zap.String("name", file.Name()))
			cnsReleaseErr := w.releaseIP(file.Name())
			if cnsReleaseErr != nil {
				logger.Error("failed to release IP from CNS", zap.Error(cnsReleaseErr))
			}
			err := RemoveFile(file.Name(), path)
			if err != nil {
				logger.Error("failed to remove file", zap.Error(err))
			}
		}
	}
	<-c.Done()
	return c.Error()
}
```

Reply:
the point of it being a blocking call is, currently we don't know if the internal goroutine is still running or has failed and exited, and if we block and use contexts to cancel we can communicate that back to the caller

Reply:
don't forget the 
```go 
			select {
			case := <-ctx.Done():
				return
```
in the select, it's important to exit the for/select loop when the outer context has been cancelled